File: Objective-C/JsonRequest/Controller/ViewController.m
//
//  ViewController.m
//  JsonRequest
//
//  Created by John Lima on 15/04/17.
//  Copyright © 2017 limadeveloper. All rights reserved.
//

#import "ViewController.h"
#import "Country.h"
#import "Global.h"

@interface ViewController () <UITableViewDataSource, UITableViewDelegate>

@end

@implementation ViewController

#pragma mark - Properties
Country *country;
NSArray *countries;
NSString *const cellName = @"cell";
UIRefreshControl *refreshControll;

#pragma mark - View LifeCycle
- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self updateUI];
    [self getData];
    [self setupRefresh];
}

#pragma mark - Actions
-(void)getData {

    [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];
    
    country = [[Country alloc] init];
    countries = [[NSArray alloc] init];
    
    [country downloadJSONFromURLWith:^(NSArray *data, NSString *error) {
        
        if (!error) {
            countries = data;
        }
        
        dispatch_async(dispatch_get_main_queue(), ^{
            
            [self updateUI];
            [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
            [refreshControll endRefreshing];
            
            if (error) {
                UIAlertAction *dismissAlert = [UIAlertAction actionWithTitle:TEXT_OK style:UIAlertActionStyleDestructive handler:nil];
                [Global showAlert:TEXT_ERROR and:error and:@[dismissAlert] and:self];
            }
        });
    }];
    
    [country downloadJSONFromURLUsingAFNetworkingWith:nil];
}

-(void)updateUI {

    NSLog(@"Countries: %d", (int)countries.count);
    
    [self.countButton setTitle:[NSString stringWithFormat:@"%d", (int)countries.count]];
    
    UIView *footer = [[UIView alloc] initWithFrame:CGRectZero];
    
    self.tableView.tableFooterView = footer;
    self.tableView.backgroundColor = [UIColor whiteColor];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    
    [self.tableView reloadData];
}

-(void)setupRefresh {

    refreshControll = [[UIRefreshControl alloc] init];
    refreshControll.tintColor = [UIColor purpleColor];
    [refreshControll addTarget:self action:@selector(refreshData) forControlEvents:UIControlEventValueChanged];
    
    [self.tableView addSubview:refreshControll];
}

-(void)refreshData {
    [self getData];
}

#pragma mark - TableView DataSource
-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return countries.count;
}

-(UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {

    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:cellName forIndexPath:indexPath];
    
    if (countries.count > 0) {
        cell.textLabel.text = [((Country *)[countries objectAtIndex:indexPath.row]) name];
    }
    
    return cell;
}

#pragma mark - TableView Delegate
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
}

@end

File: JsonRequest/ViewController.swift
//
//  ViewController.swift
//  JsonRequest
//
//  Created by John Lima on 11/18/15.
//  Copyright © 2015 John Lima. All rights reserved.
//

import UIKit

class ViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {

    // mark: properties
    @IBOutlet var tableView: UITableView!
    @IBOutlet var countries: UIBarButtonItem!
    
    var refreshControl = UIRefreshControl()
    let app = UIApplication.sharedApplication()
    
    var tableData = NSArray()
    
    // mark: view lifecycle
    override func viewDidLoad() {
        super.viewDidLoad()
        
        getData()
        setView()
        createRefresh()
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()

    }

    // mark: - actions
    func setView() {
        
        countries.title = "\(numberOfCountries())"
        
        let back = UIBarButtonItem(title: "", style: .Plain, target: nil, action: nil)
        navigationItem.backBarButtonItem = back
    }
    
    func getData() {
        app.networkActivityIndicatorVisible = true
        let json = JsonModel()
        json.downloadJSONFromURL { (json) -> () in
            dispatch_async(dispatch_get_main_queue(), { () -> Void in
                self.tableData = json!
                print("table data: \(self.tableData)")
                self.tableView.reloadData()
                self.setView()
                self.app.networkActivityIndicatorVisible = false
            })
        }
    }
    
    func createRefresh() {
        refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: "refresh:", forControlEvents: .ValueChanged)
        tableView.addSubview(refreshControl)
    }
    
    func refresh(sender:AnyObject) {
        getData()
        tableView.reloadData()
        refreshControl.endRefreshing()
    }
    
    func numberOfCountries() -> Int {
        var num = Int()
        num = tableData.count
        return num
    }
    
    // mark: - table view datasource
    func numberOfSectionsInTableView(tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return tableData.count
    }
    
    func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCellWithIdentifier("cell", forIndexPath: indexPath)
        cell.textLabel?.text = tableData.objectAtIndex(indexPath.row).valueForKey("name") as? String
        return cell
    }
    
    // mark: - table view delegate
    func tableView(tableView: UITableView, didSelectRowAtIndexPath indexPath: NSIndexPath) {
        print("index: \(indexPath.row)")
        performSegueWithIdentifier("details", sender: indexPath.row)
    }

    // mark: - segue
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        if segue.identifier == "details" {
            let index = sender as! Int
            let vc = segue.destinationViewController as! DetailViewController
            vc.tableData = [tableData.objectAtIndex(index) as! NSDictionary]
        }
    }
}

